dashboard/app/models/user.rb

```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
     if !script_sections.empty?
       # if we have one or more sections matching this script id, we consider a stage hidden if all of those sections
       # hides the stage
-      script_sections.all? {|s| script_level.stage_hidden_for_section?(s.id)}
+      script_sections.all? {|s| script_level.hidden_for_section?(s.id)}
     else
       # if we have no sections matching this script id, we consider a stage hidden if any of the sections we're in
       # hide it
-      sections.any? {|s| script_level.stage_hidden_for_section?(s.id)}
+      sections.any? {|s| script_level.hidden_for_section?(s.id)}
     end
   end
 
+  # This method will extra a list of hidden ids by section. The type of ids depends
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  def get_hidden_ids(assign_id, hidden_stages = true)
+    # If we're a teacher, we want to go through each of our sections and return
+    # a mapping from section id to hidden stages in that section
+    if teacher?
+      hidden_by_section = {}
+      sections.each do |section|
+        if hidden_stages
+          # Given a script_id, looking for hidden stages
+          hidden_by_section[section.id] = section.section_hidden_stages.pluck(:stage_id)
+        else
+          # Given a course_id, looking for hidden_scripts
+          hidden_by_section[section.id] = section.section_hidden_scripts.pluck(:script_id)
+        end
+      end
+      return hidden_by_section
+    end
+
+    # if we're a student, we want to look through each of the sections in which
+    # we're a member, and use those to figure out which stages should be hidden
+    # for us
+    sections = sections_as_student
+    return [] if sections.empty?
+
+    assigned_sections = sections.select do |section|
+      hidden_stages && section.script_id == assign_id || !hidden_stages && section.course_id == assign_id
+    end
+
+    if assigned_sections.empty?
+      # if we have no sections matching this assignment, we consider a stage/script
+      # hidden if any of our sections hides it
+      if hidden_stages
+        sections.flat_map(&:section_hidden_stages).pluck(:stage_id).uniq
+      else
+        sections.flat_map(&:section_hidden_scripts).pluck(:script_id).uniq
```

Comment:
It took me a minute to realize that lines 923 and 925 here are return values, being the last statement to execute in this path through this method even though there's 10+ lines left.  If our linter will allow it, it'd be nice to list explicit returns or assign to a local and return that at the end, to avoid confusion.

And/or, these five lines are largely duplicated with the ones in the else case below, only different by which section list we pass in and what we do with the result.  Might be a good place for a private helper.